### PR TITLE
Introduce a kernel state argument.

### PR DESCRIPTION
**Pull request by maleadt**
This PR introduces an optimization pass that adds an opaque state pointer argument (`{}* %state`) to kernel functions, and passes it to every called function so that it is available for use in GPU code. Access to this pointer is through a getter intrinsic (`{}* @julia.gpu.state_getter()`) which this optimization pass turns into a reference to the state pointer argument.

The idea is that back-ends can then cast this pointer to a backend-specific struct pointer, and pass an instance of that struct to the kernel function. The initial development goal of this PR is to be able to pass dynamic shared memory allocation info to SPIR-V kernels (which needs to be done as a function argument), but other uses are possible as well. For example, we could have it point to the exception storage location, or use it to initialize a bump allocator.

Since each thread might be accessing the state concurrently, we don't want it to be an actual pointer to global memory, so instead its being passed as a `byval` pointer. The actual type of the state is determined by the `kernel_state_type` interface method, returning a Julia type (Nothing if no type is needed).

cc @jpsamaroo @vchuravy

### CODE REVIEW

File: src/optim.jl [hunk 1]
```diff
@@ -317,3 +322,127 @@ function lower_ptls!(mod::LLVM.Module)
...
+        # clean-up
+        if f == entry
+            entry = new_f
+        end
+        @assert isempty(uses(f))
+        unsafe_delete!(mod, f)
+        LLVM.name!(new_f, fn)
+        push!(replaced, new_f)
+    end
+    empty!(worklist)
```
Comment (jpsamaroo) [inline src/optim.jl:431]: Won't the GC just delete `worklist` for us once it goes out of scope?

File: src/optim.jl [hunk 2]
```diff
@@ -317,3 +322,127 @@ function lower_ptls!(mod::LLVM.Module)
...
+                inst = user(use)
+                inst isa LLVM.CallInst || inst isa LLVM.InvokeInst || inst isa LLVM.CallBrInst || continue
+
+                # NOTE: we unconditionally add the state argument, even if there's no uses,
+                #       assuming we'll perform dead arg elimination during optimization.
+
+                # forward the state argument
+                position!(builder, inst)
+                state = call!(builder, state_getter, Value[], "state")
+                new_inst = call!(builder, new_f, [state, operands(inst)[1:end-1]...])
```
Comment (jpsamaroo) [inline src/optim.jl:414]: Shouldn't we be emitting an `InvokeInst` or `CallBrInst` here, if that was the type of `inst`?

File: src/optim.jl [hunk 3]
```diff
@@ -317,3 +322,127 @@ function lower_ptls!(mod::LLVM.Module)
...
+            entry = new_f
+        end
+        @assert isempty(uses(f))
+        unsafe_delete!(mod, f)
+        LLVM.name!(new_f, fn)
+        push!(replaced, new_f)
+    end
+    empty!(worklist)
+
+    # lower all uses of the state getter to the newly introduced function state argument
```
Comment (jpsamaroo) [inline src/optim.jl:433]:
```suggestion
    # fixup all uses of the state getter to use the newly introduced function state argument
```

**Review by jpsamaroo — APPROVED** [review]
Looks awesome!

File: src/optim.jl [hunk 4]
```diff
@@ -317,3 +322,127 @@ function lower_ptls!(mod::LLVM.Module)
...
+# `julia.gpu.state_getter` intrinsics to use this newly-introduced state argument.
+function lower_kernel_state!(mod::LLVM.Module)
+    job = current_job::CompilerJob
+    ctx = context(mod)
+
+    # find the entry-point (the only function with non-internal linkage)
+    entry = nothing
+    for f in functions(mod)
+        isdeclaration(f) && continue
+        if linkage(f) != LLVM.API.LLVMInternalLinkage
```
[the remark this answers is not in the document]
Comment (maleadt) [reply src/optim.jl:339]: We internalize all functions: https://github.com/JuliaGPU/GPUCompiler.jl/blob/abeeebf3d60c61fe1908a2d658d9219c0cb00933/src/irgen.jl#L69-L74
But yeah, I want to pass `entry` around again but that changes the API, so I want to wait until the dust settles to see what we need to change exactly.

File: src/optim.jl (hunk 1 above)
Comment (maleadt) [reply src/optim.jl:431]: Yeah, I added this after wrongfully using entries from the array (which are `unsafe_delete!`d so invalid) :smile:

File: src/optim.jl (hunk 2 above)
Comment (maleadt) [reply src/optim.jl:414]: Ah yes, good catch.

**Comment by maleadt** [comment]
Changed it to passing the state by value, so that it (at least on CUDA) can be read by each thread concurrently without synchronizing. Should also make it possible to store thread-local values in it, at least of the memory location where the state parameter is stored allows writes to it. If not, we could always extend the pass to make a copy to the stack upon entry of the kernel function. @vchuravy you'll probably be interested in this to realize https://github.com/JuliaGPU/CUDA.jl/pull/501.

**Comment by maleadt** [comment]
See https://github.com/JuliaGPU/CUDA.jl/pull/1110 for an example on how this can be used.

**Comment by jpsamaroo** [comment]
Surprisingly simple! LGTM!

**Comment by maleadt** [comment]
Yup, I'm pretty happy with how this turned out. I think it may be very useful, too!

**Review by vchuravy — APPROVED** [review]
Everybody get's some state xD

**Pull request merged**